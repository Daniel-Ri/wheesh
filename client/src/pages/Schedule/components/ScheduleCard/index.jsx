/* eslint-disable jsx-a11y/label-has-associated-control */
import { useState } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';

import { useNavigate } from 'react-router-dom';
import { formatHour, formatRupiah } from '@utils/handleValue';
import { Button } from '@mui/material';
import { createStructuredSelector } from 'reselect';
import { selectLogin } from '@containers/Client/selectors';
import { connect } from 'react-redux';
import toast from 'react-hot-toast';
import classes from './style.module.scss';

const ScheduleCard = ({ login, schedule, intl: { formatMessage } }) => {
  const navigate = useNavigate();
  const [selected, setSelected] = useState(false);

  const handleClickBook = (seatClass) => {
    // Not adding `return` so the user will be redirected to `/login`
    if (!login) {
      toast.error(formatMessage({ id: 'app_need_to_login' }));
    }

    navigate(`/book/${schedule.id}/${seatClass}`);
  };

  return (
    <div data-testid="ScheduleCard" className={`${classes.card} ${selected ? classes.selected : ''}`}>
      <div className={classes.topCard} onClick={() => setSelected((isSelected) => !isSelected)}>
        <div className={classes.leftCard}>
          <div className={classes.timeAndPlace}>
            <div className={classes.time}>
              {formatHour(schedule.departureTime)}
              <sup> WIB</sup>
            </div>
            <div className={classes.place}>{schedule.departureStation.name}</div>
          </div>
          <div className={classes.train}>
            <div className={classes.name}>{schedule.Train.name}</div>
            <div className={classes.arrowImage}>
              <img src="/arrowTrain.png" alt="" />
            </div>
          </div>
          <div className={classes.timeAndPlace}>
            <div className={classes.time}>
              {formatHour(schedule.arrivalTime)}
              <sup> WIB</sup>
            </div>
            <div className={classes.place}>{schedule.arrivalStation.name}</div>
          </div>
        </div>
        <div className={classes.price}>
          {formatRupiah(schedule.prices.filter((schedulePrice) => schedulePrice.seatClass === 'economy')[0].price)}
        </div>
      </div>
      <div className={classes.bottomCard}>
        <div className={classes.seatClass}>
          <div className={classes.nameClass}>First Class</div>
          <div className={classes.rightCard}>
            <div className={classes.priceAndAvailable}>
              <div className={classes.price}>
                {formatRupiah(schedule.prices.filter((schedulePrice) => schedulePrice.seatClass === 'first')[0].price)}
              </div>
              <div className={classes.available}>
                {/* eslint-disable-next-line no-nested-ternary */}
                {schedule.firstSeatAvailable === 'Available' ? (
                  <FormattedMessage id="app_available" />
                ) : schedule.firstSeatAvailable === 'Few' ? (
                  <FormattedMessage id="app_few" />
                ) : (
                  <FormattedMessage id="app_none" />
                )}
              </div>
            </div>
            <Button
              data-testid="BookFirstSeatButton"
              variant="contained"
              disabled={schedule.firstSeatAvailable === 'None'}
              className={classes.btn}
              onClick={() => handleClickBook('first')}
            >
              <FormattedMessage id="app_book" />
            </Button>
          </div>
        </div>
        <hr />
        <div className={classes.seatClass}>
          <div className={classes.nameClass}>Business Class</div>
          <div className={classes.rightCard}>
            <div className={classes.priceAndAvailable}>
              <div className={classes.price}>
                {formatRupiah(
                  schedule.prices.filter((schedulePrice) => schedulePrice.seatClass === 'business')[0].price
                )}
              </div>
              <div className={classes.available}>
                {/* eslint-disable-next-line no-nested-ternary */}
                {schedule.businessSeatAvailable === 'Available' ? (
                  <FormattedMessage id="app_available" />
                ) : schedule.businessSeatAvailable === 'Few' ? (
                  <FormattedMessage id="app_few" />
                ) : (
                  <FormattedMessage id="app_none" />
                )}
              </div>
            </div>
            <Button
              data-testid="BookBusinessSeatButton"
              variant="contained"
              disabled={schedule.businessSeatAvailable === 'None'}
              className={classes.btn}
              onClick={() => handleClickBook('business')}
            >
              <FormattedMessage id="app_book" />
            </Button>
          </div>
        </div>
        <hr />
        <div className={classes.seatClass}>
          <div className={classes.nameClass}>Economy Class</div>
          <div className={classes.rightCard}>
            <div className={classes.priceAndAvailable}>
              <div className={classes.price}>
                {formatRupiah(
                  schedule.prices.filter((schedulePrice) => schedulePrice.seatClass === 'economy')[0].price
                )}
              </div>
              <div className={classes.available}>
                {/* eslint-disable-next-line no-nested-ternary */}
                {schedule.economySeatAvailable === 'Available' ? (
                  <FormattedMessage id="app_available" />
                ) : schedule.economySeatAvailable === 'Few' ? (
                  <FormattedMessage id="app_few" />
                ) : (
                  <FormattedMessage id="app_none" />
                )}
              </div>
            </div>
            <Button
              data-testid="BookEconomySeatButton"
              variant="contained"
              disabled={schedule.economySeatAvailable === 'None'}
              className={classes.btn}
              onClick={() => handleClickBook('economy')}
            >
              <FormattedMessage id="app_book" />
            </Button>
          </div>
        </div>
      </div>
    </div>
  );
};

ScheduleCard.propTypes = {
  schedule: PropTypes.object.isRequired,
  login: PropTypes.bool.isRequired,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  login: selectLogin,
});

export default injectIntl(connect(mapStateToProps)(ScheduleCard));
