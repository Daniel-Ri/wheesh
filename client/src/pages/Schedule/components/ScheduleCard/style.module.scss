@import '@styles/base';

.card {
  width: 100%;
  border-radius: 10px;
  padding: 0.5rem 0.75rem;
  box-shadow: 0 0 5px 0 $color-cool-gray;
  cursor: pointer;
  transition: 0.3s;
  @include directionX;
  justify-content: space-between;

  .leftCard {
    @include directionX;
    align-items: center;
    width: 80%;
    gap: 1rem;

    .timeAndPlace {
      @include directionY;
      gap: 0.1rem;

      .time {
        font-size: 90%;

        sup {
          font-size: 60%;
          color: $color-strawberry-red;
        }
      }

      .place {
        font-size: 80%;
        color: $color-cool-gray;
      }
    }

    .train {
      min-width: 3rem;
      @include directionY;
      align-items: center;

      .name {
        margin-top: -0.75rem;
        font-size: 70%;
      }

      .arrowImage {
        margin-top: -0.9rem;
        width: 100%;
        height: 0.5rem;

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
        }
      }
    }
  }
}

.card:hover {
  box-shadow: 0 0 7px 0 $color-cool-gray;
}