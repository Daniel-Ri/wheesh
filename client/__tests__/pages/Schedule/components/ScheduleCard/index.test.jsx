import { render, fireEvent } from '@utils/testHelper';
import ScheduleCard from '@pages/Schedule/components/ScheduleCard';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

const mockProps = {
  login: false,
  schedule: {
    id: 66,
    Train: {
      id: 6,
      name: 'G1206',
    },
    departureStation: {
      id: 1,
      name: 'Halim',
    },
    arrivalStation: {
      id: 3,
      name: 'Padalarang',
    },
    departureTime: '2023-12-15T02:45:00.000Z',
    arrivalTime: '2023-12-15T03:15:00.000Z',
    firstSeatAvailable: 'Available',
    businessSeatAvailable: 'Available',
    economySeatAvailable: 'Available',
    prices: [
      {
        id: 196,
        scheduleId: 66,
        seatClass: 'economy',
        price: 200000,
      },
      {
        id: 197,
        scheduleId: 66,
        seatClass: 'business',
        price: 450000,
      },
      {
        id: 198,
        scheduleId: 66,
        seatClass: 'first',
        price: 600000,
      },
    ],
  },
};

beforeEach(() => {
  wrapper = render(<ScheduleCard {...mockProps} />);
});

describe('Component ScheduleCard of Schedule page', () => {
  test('Renders ScheduleCard component', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('ScheduleCard')).toBeInTheDocument();
  });

  test("Should navigate when click 'Book' button", () => {
    const { getByTestId } = wrapper;
    const bookEconomySeatButton = getByTestId('BookEconomySeatButton');
    fireEvent.click(bookEconomySeatButton);
    expect(mockNavigate).toHaveBeenCalledWith(`/book/${mockProps.schedule.id}/economy`);
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
