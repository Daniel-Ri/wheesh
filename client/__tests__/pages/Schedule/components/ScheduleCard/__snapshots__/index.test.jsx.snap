// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ScheduleCard of Schedule page Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card "
        data-testid="ScheduleCard"
      >
        <div
          class="topCard"
        >
          <div
            class="leftCard"
          >
            <div
              class="timeAndPlace"
            >
              <div
                class="time"
              >
                09:45
                <sup>
                   WIB
                </sup>
              </div>
              <div
                class="place"
              >
                Halim
              </div>
            </div>
            <div
              class="train"
            >
              <div
                class="name"
              >
                G1206
              </div>
              <div
                class="arrowImage"
              >
                <img
                  alt=""
                  src="/arrowTrain.png"
                />
              </div>
            </div>
            <div
              class="timeAndPlace"
            >
              <div
                class="time"
              >
                10:15
                <sup>
                   WIB
                </sup>
              </div>
              <div
                class="place"
              >
                Padalarang
              </div>
            </div>
          </div>
          <div
            class="price"
          >
            Rp 200.000
          </div>
        </div>
        <div
          class="bottomCard"
        >
          <div
            class="seatClass"
          >
            <div
              class="nameClass"
            >
              First Class
            </div>
            <div
              class="rightCard"
            >
              <div
                class="priceAndAvailable"
              >
                <div
                  class="price"
                >
                  Rp 600.000
                </div>
                <div
                  class="available"
                >
                  Tersedia
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="BookFirstSeatButton"
                tabindex="0"
                type="button"
              >
                Pesan
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <hr />
          <div
            class="seatClass"
          >
            <div
              class="nameClass"
            >
              Business Class
            </div>
            <div
              class="rightCard"
            >
              <div
                class="priceAndAvailable"
              >
                <div
                  class="price"
                >
                  Rp 450.000
                </div>
                <div
                  class="available"
                >
                  Tersedia
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="BookBusinessSeatButton"
                tabindex="0"
                type="button"
              >
                Pesan
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <hr />
          <div
            class="seatClass"
          >
            <div
              class="nameClass"
            >
              Economy Class
            </div>
            <div
              class="rightCard"
            >
              <div
                class="priceAndAvailable"
              >
                <div
                  class="price"
                >
                  Rp 200.000
                </div>
                <div
                  class="available"
                >
                  Tersedia
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="BookEconomySeatButton"
                tabindex="0"
                type="button"
              >
                Pesan
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card "
      data-testid="ScheduleCard"
    >
      <div
        class="topCard"
      >
        <div
          class="leftCard"
        >
          <div
            class="timeAndPlace"
          >
            <div
              class="time"
            >
              09:45
              <sup>
                 WIB
              </sup>
            </div>
            <div
              class="place"
            >
              Halim
            </div>
          </div>
          <div
            class="train"
          >
            <div
              class="name"
            >
              G1206
            </div>
            <div
              class="arrowImage"
            >
              <img
                alt=""
                src="/arrowTrain.png"
              />
            </div>
          </div>
          <div
            class="timeAndPlace"
          >
            <div
              class="time"
            >
              10:15
              <sup>
                 WIB
              </sup>
            </div>
            <div
              class="place"
            >
              Padalarang
            </div>
          </div>
        </div>
        <div
          class="price"
        >
          Rp 200.000
        </div>
      </div>
      <div
        class="bottomCard"
      >
        <div
          class="seatClass"
        >
          <div
            class="nameClass"
          >
            First Class
          </div>
          <div
            class="rightCard"
          >
            <div
              class="priceAndAvailable"
            >
              <div
                class="price"
              >
                Rp 600.000
              </div>
              <div
                class="available"
              >
                Tersedia
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="BookFirstSeatButton"
              tabindex="0"
              type="button"
            >
              Pesan
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <hr />
        <div
          class="seatClass"
        >
          <div
            class="nameClass"
          >
            Business Class
          </div>
          <div
            class="rightCard"
          >
            <div
              class="priceAndAvailable"
            >
              <div
                class="price"
              >
                Rp 450.000
              </div>
              <div
                class="available"
              >
                Tersedia
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="BookBusinessSeatButton"
              tabindex="0"
              type="button"
            >
              Pesan
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <hr />
        <div
          class="seatClass"
        >
          <div
            class="nameClass"
          >
            Economy Class
          </div>
          <div
            class="rightCard"
          >
            <div
              class="priceAndAvailable"
            >
              <div
                class="price"
              >
                Rp 200.000
              </div>
              <div
                class="available"
              >
                Tersedia
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="BookEconomySeatButton"
              tabindex="0"
              type="button"
            >
              Pesan
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
